File: scripts/validate.js
#!/usr/bin/env node

/**
 * Validate quotes/quotes.json schema and catch common issues.
 */

const fs = require('fs');
const path = require('path');

const QUOTES_PATH = path.join(__dirname, '..', 'quotes', 'quotes.json');

const VALID_CATEGORIES = [
  'grug', 'suffering', 'indie-hacker', 'unix', 'debugging',
  'complexity', 'shipping', 'career', 'wisdom', 'shitpost',
  'founder', 'tools', 'funny'
];

function validate() {
  const quotes = JSON.parse(fs.readFileSync(QUOTES_PATH, 'utf-8'));
  const errors = [];
  const warnings = [];
  const ids = new Set();

  for (let i = 0; i < quotes.length; i++) {
    const q = quotes[i];
    const prefix = `Quote #${i} (${q.id || 'NO ID'})`;

    // Required fields
    if (!q.id) errors.push(`${prefix}: missing 'id'`);
    if (!q.text) errors.push(`${prefix}: missing 'text'`);
    if (!q.author) errors.push(`${prefix}: missing 'author'`);
    if (!q.author_slug) errors.push(`${prefix}: missing 'author_slug'`);
    if (!q.categories || !q.categories.length) errors.push(`${prefix}: missing 'categories'`);
    if (!q.tags || !q.tags.length) errors.push(`${prefix}: missing 'tags'`);
    if (q.grug_level === undefined) errors.push(`${prefix}: missing 'grug_level'`);
    if (q.spiciness === undefined) errors.push(`${prefix}: missing 'spiciness'`);

    // Duplicate ID check
    if (q.id && ids.has(q.id)) errors.push(`${prefix}: duplicate ID`);
    if (q.id) ids.add(q.id);

    // Range checks
    if (q.grug_level < 1 || q.grug_level > 5) errors.push(`${prefix}: grug_level must be 1-5, got ${q.grug_level}`);
    if (q.spiciness < 1 || q.spiciness > 5) errors.push(`${prefix}: spiciness must be 1-5, got ${q.spiciness}`);

    // Category validation
    if (q.categories) {
      for (const cat of q.categories) {
        if (!VALID_CATEGORIES.includes(cat)) {
          warnings.push(`${prefix}: unknown category '${cat}' (valid: ${VALID_CATEGORIES.join(', ')})`);
        }
      }
    }

    // Slug format check
    if (q.id && !/^[a-z0-9-]+$/.test(q.id)) {
      errors.push(`${prefix}: id should be lowercase with hyphens only`);
    }
    if (q.author_slug && !/^[a-z0-9-]+$/.test(q.author_slug)) {
      errors.push(`${prefix}: author_slug should be lowercase with hyphens only`);
    }
  }

  // Report
  console.log(`\n📋 Validated ${quotes.length} quotes\n`);

  if (warnings.length) {
    console.log(`⚠️  ${warnings.length} warnings:`);
    warnings.forEach(w => console.log(`   ${w}`));
    console.log();
  }

  if (errors.length) {
    console.log(`❌ ${errors.length} errors:`);
    errors.forEach(e => console.log(`   ${e}`));
    process.exit(1);
  } else {
    console.log('✅ All quotes valid!');
    console.log(`   ${ids.size} unique quotes`);
    console.log(`   ${new Set(quotes.map(q => q.author_slug)).size} unique authors`);
    console.log(`   ${new Set(quotes.flatMap(q => q.categories)).size} categories used`);
    console.log(`   ${new Set(quotes.flatMap(q => q.tags)).size} unique tags`);
  }
}

validate();

File: scripts/validate.py
#!/usr/bin/env python3
"""
Validate quotes/quotes.json schema and catch common issues.
Run via: python3 scripts/validate.py
"""

import json
import re
import sys
from pathlib import Path

QUOTES_PATH = Path(__file__).parent.parent / 'quotes' / 'quotes.json'

VALID_CATEGORIES = [
    'grug', 'suffering', 'indie-hacker', 'unix', 'debugging',
    'complexity', 'shipping', 'career', 'wisdom', 'shitpost',
    'founder', 'tools', 'funny', 'design', 'product', 'engineering',
    'business', 'entrepreneur', 'marketing', 'saas', 'users',
    'productivity', 'devops'
]


def validate():
    with open(QUOTES_PATH, encoding='utf-8') as f:
        quotes = json.load(f)

    errors = []
    warnings = []
    ids = set()

    for i, q in enumerate(quotes):
        prefix = f"Quote #{i} ({q.get('id', 'NO ID')})"

        # Required fields
        for field in ['id', 'text', 'author', 'author_slug', 'categories', 'tags']:
            if not q.get(field):
                errors.append(f"{prefix}: missing '{field}'")

        if q.get('grug_level') is None:
            errors.append(f"{prefix}: missing 'grug_level'")
        if q.get('spiciness') is None:
            errors.append(f"{prefix}: missing 'spiciness'")

        # Duplicate ID check
        qid = q.get('id')
        if qid:
            if qid in ids:
                errors.append(f"{prefix}: duplicate ID")
            ids.add(qid)

        # Range checks
        gl = q.get('grug_level', 0)
        sp = q.get('spiciness', 0)
        if not (1 <= gl <= 5):
            errors.append(f"{prefix}: grug_level must be 1-5, got {gl}")
        if not (1 <= sp <= 5):
            errors.append(f"{prefix}: spiciness must be 1-5, got {sp}")

        # Category validation
        for cat in q.get('categories', []):
            if cat not in VALID_CATEGORIES:
                warnings.append(f"{prefix}: unknown category '{cat}'")

        # Slug format
        if qid and not re.match(r'^[a-z0-9-]+$', qid):
            errors.append(f"{prefix}: id should be lowercase with hyphens only")
        slug = q.get('author_slug', '')
        if slug and not re.match(r'^[a-z0-9-]+$', slug):
            errors.append(f"{prefix}: author_slug should be lowercase with hyphens only")

    # Report
    print(f"\n📋 Validated {len(quotes)} quotes\n")

    if warnings:
        print(f"⚠️  {len(warnings)} warnings:")
        for w in warnings:
            print(f"   {w}")
        print()

    if errors:
        print(f"❌ {len(errors)} errors:")
        for e in errors:
            print(f"   {e}")
        sys.exit(1)
    else:
        all_cats = set(c for q in quotes for c in q.get('categories', []))
        all_tags = set(t for q in quotes for t in q.get('tags', []))
        all_authors = set(q['author_slug'] for q in quotes)
        print('✅ All quotes valid!')
        print(f'   {len(ids)} unique quotes')
        print(f'   {len(all_authors)} unique authors')
        print(f'   {len(all_cats)} categories used')
        print(f'   {len(all_tags)} unique tags')


if __name__ == '__main__':
    validate()
